graph/src/components/store/entity_cache.rs

```diff
@@ -203,18 +202,94 @@ impl EntityCache {
 
         let query = DerivedEntityQuery {
             entity_type: EntityType::new(base_type.to_string()),
-            entity_field: field.name.clone().to_snake_case().into(),
+            entity_field: field.name.clone().into(),
             value: eref.entity_id.clone(),
             causality_region: eref.causality_region,
             id_is_bytes: id_is_bytes,
         };
 
-        let entities = self.store.get_derived(&query)?;
-        entities.iter().for_each(|(key, e)| {
-            self.current.insert(key.clone(), Some(e.clone()));
-        });
-        let entities: Vec<Entity> = entities.values().cloned().collect();
-        Ok(entities)
+        let mut entity_map = self.store.get_derived(&query)?;
+
+        for (key, entity) in entity_map.iter() {
+            // Only insert to the cache if it's not already there
+            // This is to avoid overwriting updates
+            if !self.current.contains_key(&key) {
+                self.current.insert(key.clone(), Some(entity.clone()));
+            }
+        }
+
+        // Apply updates from `updates` and `handler_updates` directly to entities in `entity_map` that match the query
+        for (key, entity) in entity_map.iter_mut() {
+            let mut entity_cow = Some(Cow::Borrowed(entity));
+
+            if let Some(op) = self.updates.get(key).cloned() {
+                op.apply_to(&mut entity_cow)
+                    .map_err(|e| key.unknown_attribute(e))?;
+            }
+
+            if let Some(op) = self.handler_updates.get(key).cloned() {
+                op.apply_to(&mut entity_cow)
+                    .map_err(|e| key.unknown_attribute(e))?;
+            }
+
+            if let Some(updated_entity) = entity_cow {
+                *entity = updated_entity.into_owned();
+            }
+        }
+
+        // A helper function to apply an update and return the resulting entity if it matches the query
+        fn apply_update(
+            op: &EntityOp,
+            query: &DerivedEntityQuery,
+            key: &EntityKey,
+        ) -> Result<Option<Entity>, anyhow::Error> {
+            match op {
+                EntityOp::Update(entity) | EntityOp::Overwrite(entity)
+                    if query.matches(key, entity) =>
+                {
+                    Ok(Some(entity.clone()))
+                }
+                EntityOp::Remove => Ok(None),
+                _ => Ok(None),
+            }
+        }
```

Comment:
I think the comment and the name of the function are what's confusing me - we are not applying an update. We use this only on entities that do not exist in the store. That means this is either the creation of an entity or the deletion of a nonexistent entity. We can ignore the deletion, and for the creation we want to add it to `entity_map` assuming it matches the query.

Maybe `is_matching_creation` or `matches_query` would be a better name for this function. And the comment should also be changed.